### ISSUE TITLE
Slow performance under load

### ISSUE THREAD

**Issue by rmsaksida** [learningtapestry/analytics#289]
We should look into doing some performance optimizations for the project.

Ref. @shlag3n.

```
Each request should take less than 150ms, now it's taking up to 16s (usually 2-5s)
Number of tracked users: ~14 000 and it's increasing

Should process
1. Up to 16 concurrent requests
2. Request type: site details or page details for up to 150 users (asking in batches for 20 users at once) for the last 3-4 days (most requests, but sometimes for last month) for one site domain.

App specific: the most load is on weekdays, 7am-9am EST.

querying analytics via an api (actually with analytics gem) and it's start page for most of users.

gem is using api, sites details, page details from api for 1 domain and up to 20 users per call
```

**Comment by hungryzi** [comment]
I discussed this further with @shlag3n.

This is how they're using the [`analytics` gem](https://github.com/learningtapestry/analytics-ruby-client):

```
lt_agent.add_filter :site_domains, [site_domain]
lt_agent.usernames = user_ids.map { |id| course.lt_user_id(id) }
lt_agent.entity = entity
lt_agent.type = type
lt_agent.add_filter :date_begin, date_begin
lt_agent.add_filter :date_end, date_end
```

> this is how we're using analytics-ruby-client, `entity='page_visits', type='details'`, `date_end=now, date_begin 3 up to 1 month from date_end`
> we're using it to show stats of how students (school students) are using online courses. teachers and school administrators are using it on daily basis and we're showing these stats on front page.
> i think we do need performance tests at least for 2 types of requests (better to start from `page details`) for 1 domain and up to 20 users per call. db need to be populated with ~30 000 users with history for ~1year usage (1-40 records per day)
> rm did some changes and performance became better, from 5-6 seconds to 100-200ms. but sometimes it's still around ~1s. too much for these requests.

@rm I'm thinking about writing some stress test where I would insert a large amount of data (as described above) and test that the API returns the results under a time limit. This would be my first time writing stress test, do you have any suggestions?

**Comment by rmsaksida** [comment]
@hungryzi Sounds good. Some general tips for that:

- You'll be able to populate the DB faster by writing raw SQL or using [activerecord-import](https://github.com/zdennis/activerecord-import). Inserting objects using plain ActiveRecord is very slow. Of course, you will only have to do it once, so it might be worth just waiting a few minutes instead of doing it in a more efficient way.
- JMeter works fine for me for stress testing.

Also, kind of obvious, but we don't know what the bottleneck is exactly. So instead of starting from a fully populated DB and multiple requests, you could check whether the response times are sufficient for a single request and/or smaller loads, maybe there's an obvious optimization we can do that will scale to higher loads.

Your computer is probably faster than the production server, so maybe it'd better to use staging for testing solutions.

**Comment by hungryzi** [comment]
@rm good points about bottleneck. I will definitely poke around for low hanging fruits.

Thanks to the recommendation of JMeter. I've just downloaded and will try it out.

**Comment by hungryzi** [comment]
Ref #291

TLDR; I propose a few adjustments:

1. Increase the `work_mem` for Postgres from `4MB` to `40MB`.
1. Add 2 more indices.
1. Use the more performant `Oj.dump` instead of the default `to_json`.
1. Read the results as arrays instead of initialize ActiveRecord object.

### Details:

#### Increase `work_mem`

I started by using `EXPLAIN ANALYZE` with the slow query. This is a result on my computer:

```
QUERY PLAN
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 GroupAggregate  (cost=336.30..344.66 rows=304 width=239) (actual time=6672.334..7552.296 rows=1350 loops=1)
   Group Key: sites.url, sites.display_name, pages.display_name, pages.url, users.username
   ->  Sort  (cost=336.30..337.06 rows=304 width=239) (actual time=6671.691..7514.948 rows=105600 loops=1)
         Sort Key: sites.url, sites.display_name, pages.display_name, pages.url, users.username
         **Sort Method: external merge  Disk: 7016kB**
         ->  Hash Left Join  (cost=12.19..323.76 rows=304 width=239) (actual time=0.486..314.579 rows=105600 loops=1)
               Hash Cond: (pages.site_id = sites.id)
               ->  Nested Loop Left Join  (cost=0.70..308.09 rows=304 width=194) (actual time=0.292..276.473 rows=105600 loops=1)
                     ->  Nested Loop  (cost=0.42..215.11 rows=304 width=67) (actual time=0.288..79.496 rows=105600 loops=1)
                           ->  Seq Scan on users  (cost=0.00..40.38 rows=1 width=63) (actual time=0.250..0.292 rows=30 loops=1)
                                 Filter: (organization_id = 26)
                                 Rows Removed by Filter: 1374
                           ->  Index Scan using index_visits_on_user_id on visits  (cost=0.42..171.06 rows=368 width=12) (actual time=0.033..1.965 rows=3520 loops=30)
                                 Index Cond: (user_id = users.id)
                                 Filter: ((date_visited >= '2016-04-15 09:31:10.239'::timestamp without time zone) AND (date_visited <= '2017-04-15 09:31:10.239'::timestamp without time zone))
                                 Rows Removed by Filter: 120
                     ->  Index Scan using pages_pkey on pages  (cost=0.28..0.30 rows=1 width=135) (actual time=0.001..0.001 rows=1 loops=105600)
                           Index Cond: (id = visits.page_id)
               ->  Hash  (cost=7.33..7.33 rows=333 width=53) (actual time=0.180..0.180 rows=261 loops=1)
                     Buckets: 1024  Batches: 1  Memory Usage: 14kB
                     ->  Seq Scan on sites  (cost=0.00..7.33 rows=333 width=53) (actual time=0.002..0.109 rows=261 loops=1)
 Planning time: 1.417 ms
 Execution time: 7576.593 ms
```

Notice how the Sort involved an external merge. It looks like it run out of memory. I checked and my `work_mem` was `4MB`. Increasing to `20MB` allows sort to use quicksort in memory:

```
->  Sort  (cost=4797.80..4820.42 rows=9048 width=232) (actual time=3386.911..3401.443 rows=105600 loops=1)
         Sort Key: sites.url, sites.display_name, pages.display_name, pages.url, users.username
         Sort Method: quicksort  Memory: 17348kB
```

This cuts the query time in half on my computer. Staging seems to use `4MB` too, I'm not sure about production.

#### Indices

Adding the `org_api_key` index on `organizations` speeds up the other query on that request.

Adding the `organization_id` index on `users` helps with the join, but the difference isn't significant.

#### Using OJ gem

After optimizing the query time, I focused on optimizing generating the JSON response. Switching to OJ gem reduces request time, especially for request with a lot of results.

#### Read the results as arrays instead of initialize ActiveRecord objects.

I experimented with this and it significantly reduces the response time for requests with a lot of results.

**Comment by rmsaksida** [comment]
Thank you for the through research, this looks fantastic.

**Comment by rmsaksida** [comment]
@shlag3n the changes have been applied to staging and I've also updated it to use more `work_mem`. Can you please check whether API response times have improved? Not sure staging is helpful with that.

**Comment by shlag3n** [comment]
Thank you, i can't say re performance bc we don't have newrelic for staging. I can populate lt-analytics staging with fake data and test it but imho it'll be double work and we can rely on performance tests results here - do we have avg time response for 15 000 users with daily activity?

More important that response format was changed, total_time is supposed to be numeric type (check swagger spec) and now it's string. We do need to decide where changes need to be done - at analytics or at edhesive (and possibly other analytic's clients) side.
Please, don't deploy it on production before that bc current edhesive code will handle it like an error.

**Comment by rmsaksida** [comment]
@hungryzi can you check regarding the datatype change for the `total_time` property?

**Comment by hungryzi** [comment]
@shlag3n Good catch! I have changed it back to Integer. I've generated data on staging. You can test it with the request below (i'm not sure if I'm overdoing it but this request returns 20MB JSON and kinda kills my Chrome tab so beware; You can of course reduce the time range). Server took 10s to start responding.

http://pelican.analytics-606785.staging.c66.me/api/v2/pages?org_api_key=9dc2774c-9e0a-4f97-ae3d-ef0e809ae980&org_secret_key=dd894d52-e9e6-4ea2-aebe-5aa55e089e7b&usernames=vicenta,luther,amiee_daniel,eladia_becker,gladis,nikole_pouros,mora.pagac,verona,liberty,jacelyn.towne,bernardo,sandee_grant,ava.conroy,alexandra_jacobi,hee_lynch,myrta,stephen.schmitt,maurine,lincoln,lida_bauch,malik,inocencia_hartmann,garland_robel,shandi,beulah.gorczany,curtis.king,gwyn.bahringer,aracelis.zieme,lilian_okon,caitlin&date_begin=2017-01-19T09:31:10.239Z&date_end=2017-04-19T09:31:10.239Z&type=detail

If you wanna try it out on your local machine, you can run `rake generate:visits` to generate the sample data.

Also, how did you generate the `swagger.yml` doc for the API? I'd like to try it too.

**Comment by shlag3n** [comment]
@rmsaksida, please, let me know when it'll be deployed on staging. i'll test it again and let you know if it's ok to go to production (at least from edhesive side).

@hungryzi, thank you, 10s for this heavy request sounds very good. Actually i meant that it'd be good to have performance autotests that will measure if we meet or not requirements on every build. Anyway it's up to you, @edgarf and @rmsaksida  how to maintain this project. I mean you need to discuss with them if this task is completed or not.
I'll check performance difference based on newrelic reports when it'll be deployed on production, but as you understand it'll be one-time manual check and with no guarantee re possible future changes. Edhesive is using analytics as 3rd party service and expecting some kind of SLA is normal, but i believe it's @edgarf question.

@hungryzi, re swagger - i don't know, better to ask rm or at slack channel.

**Comment by hungryzi** [comment]
@shlag3n good point on the performance regression test. @rmsaksida & @edgarf should I add a performance test into our suite? It will increase our test running time by 1-2 minutes, which I assume should be fine.

**Comment by rmsaksida** [comment]
@shlag3n @hungryzi yes, sounds good.

@shlag3n I'm deploying to staging.

**Comment by shlag3n** [comment]
@rmsaksida, @hungryzi - tested with edhesive. thank you, looks good now.
